test: Enable SC2046 and SC2086 shellcheck rules

hebasto opened the pull request:
Closes #20879.
Replaces #22695.

**Note for reviewers**. Some touched shell scripts are not being run in CI, therefore they require more thorough reviewing:
- `contrib/devtools/gen-manpages.sh`
- `contrib/macdeploy/detached-sig-apply.sh`
- `contrib/windeploy/detached-sig-create.sh`
- `src/qt/res/animation/makespinner.sh`

hebasto committed:
test: Enable SC2086 shellcheck rule

hebasto committed:
test: Enable SC2046 shellcheck rule

dougEfresh commented:
Since you are already touching the ci scripts with double quotes  `"` 
Maybe add `"` around the  `cd P_CI_DIR` in  DOCKER_EXEC function: 

```
DOCKER_CI_CMD_PREFIX bash -c "export PATH=$BASE_SCRATCH_DIR/bins/:\$PATH && cd \"$P_CI_DIR\" && $*"
```

hebasto commented:
Friendly ping @dongcarl and @practicalswift 
:tiger2:

Zero-1729 commented:
Approach ACK

MarcoFalke commented on ci/test/00_setup_env_native_qt5.sh:
```diff
@@ -15,5 +15,5 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
 export GOAL="install"
 export PREVIOUS_RELEASES_TO_DOWNLOAD="v0.15.2 v0.16.3 v0.17.2 v0.18.1 v0.19.1 v0.20.1"
-export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-reduce-exports
+export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-reduce-exports \
```
Does this change change behavior?

hebasto replied:
No, it doesn't. Why do you think so?

MarcoFalke replied:
Ok, was just wondering if there is a difference.

hebasto replied:
More details.

Being quoted, the `"$BITCOIN_CONFIG` preserves the `newline` character.

Now it is [treated](https://www.gnu.org/software/bash/manual/html_node/Escape-Character.html#Escape-Character) as a line continuation.

hebasto commented:
Updated 02e9579cecc8002e37013dc11a0b7ab5b7a90d42 -> 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e ([pr23462.03](https://github.com/hebasto/bitcoin/commits/pr23462.03) -> [pr23462.04](https://github.com/hebasto/bitcoin/commits/pr23462.04)):

- rebased due to the conflict with #23458
- addressed @dougEfresh's [comment](https://github.com/bitcoin/bitcoin/pull/23462#issuecomment-962671897)

laanwj commented:
Tested successfully:
- `contrib/devtools/gen-manpages.sh`
- `src/qt/res/animation/makespinner.sh`

ifuckingyouyeah commented:
![download](https://user-images.githubusercontent.com/93928639/140761429-f4a24c36-70a7-4429-a724-198357851cc3.jpeg)
kiss my dick bitch

Zero-1729 reviewed (commented):
tACK 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e

Tested on macOS v12.0.1 (21A559)

Successfully tested the following files locally and can verify the SC2046 & SC2086 warnings fixed:

> **Note**: On macOS, the default `sed` triggers an error when running `gen-manpages.sh`. This is resolved by using the GNU version `gnu-sed` instead (i.e. install and prepend its path in `PATH` to override the default `sed`). Running `src/qt/res/animation/makespinner.sh` also requires `imagemagick`. `contrib/macdeploy/detached-sig-apply.sh` requires [signapple](https://github.com/achow101/signapple) installed.

- contrib/devtools/gen-manpages.sh
- src/qt/res/animation/makespinner.sh
- contrib/macdeploy/detached-sig-apply.sh

**PS**: the minor notes above are intended only for those who haven't run these files before on macOS.

laanwj commented:
> Note: On macOS, the default sed triggers an error when running gen-manpages.sh.

Wrt general gen-manpages.sh frustrations: please see #22681. The sed parts are likely going to be removed completely before 0.23.

laanwj commented on ci/test/04_install.sh:
```diff
@@ -12,6 +12,7 @@ fi
 
 if [ "$CI_OS_NAME" == "macos" ]; then
   sudo -H pip3 install --upgrade pip
+  # shellcheck disable=SC2086
```
What is the idea of disabling these specific checks in some places? Is the idea that these are temporary and need to be fixed later? (but too much hassle to fix in thir PS?)

hebasto replied:
> What is the idea of disabling these specific checks in some places?

```bash
$ PIP_PACKAGES="zmq lief"
$ pip3 install --user "$PIP_PACKAGES"
ERROR: Invalid requirement: 'zmq lief'
```

hebasto replied:
Looks like that. And that code requires `# shellcheck disable=SC2086`.

hebasto replied:
Also see https://github.com/koalaman/shellcheck/wiki/SC2086#exceptions

hebasto replied:
This works:
```bash
$ PIP_PACKAGES=(zmq lief)
$ pip3 install --user "${PIP_PACKAGES[@]}"
```

laanwj replied:
Using arrays sounds like a good idea to me. At least for scripts like this that explicitly need `bash`.

dongcarl commented:
Concept ACK, unintentional globbing and word splitting lead to inscrutable bugs.

We should mark with a `disable=` wherever we need them or simply use arrays.

hebasto commented:
> Using arrays sounds like a good idea to me. At least for scripts like this that explicitly need `bash`.

> We should mark with a `disable=` wherever we need them or simply use arrays.

Could we leave introducing arrays for a follow up?

laanwj commented:
> Could we leave introducing arrays for a follow up?

Fine with me. Merging this soon closes off new such issues being introduced. And it's already the second try.

hebasto commented:
Updated 27fbb10cbc5fb82e52e5b62dc34cd5c5c729452e -> fe0ff569ea6c353f88609c0f5f9b6fa80ff74f15 ([pr23462.04](https://github.com/hebasto/bitcoin/commits/pr23462.04) -> [pr23462.05](https://github.com/hebasto/bitcoin/commits/pr23462.05)):

- rebased due to the conflict with #23114
- the second commit has been re-worked, and discovered bugs have been fixed

> What is the idea of disabling these specific checks in some places?

Now this PR introduces `# shellcheck disable=SC2046` in two places only:
- `contrib/windeploy/detached-sig-create.sh` is not a `bash` script, that is required to use arrays
- in `test/lint/lint-python.sh` too much hassle is required to avoid it

laanwj commented:
Code review re-ACK fe0ff569ea6c353f88609c0f5f9b6fa80ff74f15

fanquake commented:
Possible this broke the verify commits pre push hook. See https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/198 / https://github.com/bitcoin/bitcoin/pull/35235.
